agent/lib/kontena/observer.rb

```diff
@@ -1,117 +1,306 @@
+    #
+    # @raise [Timeout::Error]
+    def receive
+      timeout = @deadline ? @deadline - Time.now : nil
+
+      debug { "receive timeout=#{timeout}..." }
 
-        debug "observe Observable<#{observable.__klass__}> -> #{value}"
+      begin
+        # Celluloid.receive => Celluloid::Actor#receive => Celluloid::Internals::Receiver#receive returns nil on timeout
+        message = Celluloid.receive(timeout) { |msg| Kontena::Observable::Message === msg && msg.observer == self }
+      rescue Celluloid::TaskTimeout
+        # Celluloid.receive => Celluloid::Mailbox.receive raises TaskTimeout insstead
+        message = nil
+      end
 
-        # crash if observed Actor crashes, otherwise we get stuck without updates
-        # this is not a bidrectional link: our crashes do not propagate to the observable
-        self.monitor observable
+      if message
+        return message
+      else
+        raise Timeout::Error, "observe timeout #{'%.2fs' % timeout}: #{self.describe_observables}"
       end
+    end
+
+    # Any observable has an error?
+    #
+    # @return [Boolean] true => call to crash
+    def error?
+      @values.any? { |observable, value| Exception === value }
+    end
+
+    # Each observable has a value?
+    #
+    # @return [Boolean] false => block calls on updates
+    def ready?
+      !@values.any? { |observable, value| value.nil? }
+    end
+
+    # Map each observable to its value
+    #
+    # @return [Array] values or nil
+    def values
+      @observables.map{|observable| @values[observable] }
+    end
 
-      # immediate async update if all observables were ready
-      async.observed(observe)
+    # @return [Exception, nil]
+    def error
+      @values.each_pair{|observable, value|
+        return Error.new(observable, value) if Exception === value
+      }
+      return nil
+    end
+
+    # Yield each set of ready? observed values while alive, or raise on error?
+    #
+    # The yield is exclusive, because suspending the observing task would mean that
+    # any observable messages would get discarded.
+    #
+    # @param timeout [Float] timeout between each yield
+    def each(timeout: nil)
+      @deadline = Time.now + timeout if timeout
 
-      observe
+      while alive?
+        # prevent any intervening messages from being processed and discarded before we're back in Celluloid.receive()
+        Celluloid.exclusive {
+          if error?
+            debug { "raise: #{self.describe_observables}" }
+
+            raise self.error
+          elsif ready?
+            debug { "yield: #{self.describe_observables}" }
+
+            yield *self.values
+
+            @deadline = Time.now + timeout if timeout
+          else
+            debug { "wait: #{self.describe_observables}" }
+          end
+        }
+
+        # must be atomic!
+        update(receive())
+      end
+    end
+
+    # No longer expecting updates.
+    #
+    def kill
+      @alive = false
```

Comment:
It's not the intention to be able to "kill" on observer that's receiving. This is only intended to be called from the observing task itself, and that only happens when the observe call returns (per `ensure`). The observing task can't be receiving then. Crashing an observable will wake up and kill the observers.

Maybe this is badly named. The `Kontena::Observe.observe` interface doesn't expose it, but there might be some usecase for using `Kontena::Observer.new `/ `observer.observe` / `observe.each` separately, and in that case, the caller is responsible for calling `observer.kill` to not leak the observer. If the observer stops receiving, then the observed observables will still have that observer in their observers hash, and will continue sending messages to the observing actor's mailbox that just get discarded.

Reply:
The `while alive?` loop in the `Observer#each` method is kinda cosmetic... there is no known case where that would ever be false. It could be changed to `while true`.